File: pgmq/Producer_test.go
package pgmq_test

import (
	"context"
	"database/sql"
	"testing"

	"github.com/rizvn/pgutil/pgmq"
	"github.com/rizvn/pgutil/testutil"
)

func TestProducer(t *testing.T) {
	// Start Postgres test container
	ctr, dsn, err := testutil.StartPgTestContainer()
	if err != nil {
		t.Fatal(err)
	}
	defer func() { _ = ctr.Terminate(context.Background()) }()

	// Create db pool
	dbPool, err := sql.Open("pgx", dsn)
	if err != nil {
		t.Fatalf("failed to create db p pool pool %v", err)
	}
	dbPool.SetMaxOpenConns(20)

	// Create pgmq queue
	_, err = dbPool.Exec(`SELECT *  FROM pgmq.create('test_queue')`)
	if err != nil {
		t.Fatalf("failed to create pgmq queue: %v", err)
	}

	// Clean up pgmq queue so its empty
	_, err = dbPool.Exec(`SELECT * FROM pgmq.purge_queue('test_queue')`)
	if err != nil {
		t.Fatalf("failed to truncate pgmq queue: %v", err)
	}

	// Act

	// Create producer
	p := pgmq.NewProducer(dbPool)

	// Produce a message
	err = p.Produce("test_queue", `{"content": "Hello, World!"}`, "{}")
	if err != nil {
		t.Fatal(err)
	}

	// Assert - check if message was produced in pgmq table
	row, err := dbPool.Query(`SELECT count(0) FROM pgmq.q_test_queue`)
	if err != nil {
		t.Fatalf("failed to query pgmq queue: %v", err)
	}

	var msgCount int
	for row.Next() {
		err = row.Scan(&msgCount)
		if err != nil {
			t.Fatalf("failed to scan message count: %v", err)
		}
	}

	if msgCount == 0 {
		t.Errorf("Expected to have produced messages, but message count is 0")
	}

}

File: src/test/java/com/github/rizvn/pgmq/ProducerTest.java
package com.github.rizvn.pgmq;

import com.github.rizvn.DbTest;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import java.sql.Connection;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.sql.Statement;

import static org.junit.jupiter.api.Assertions.assertNotEquals;
import static org.junit.jupiter.api.Assertions.assertTrue;

class ProducerTest extends DbTest {

    private Producer producer;
    private static final String QUEUE_NAME = "test_queue";

    @BeforeEach
    void setUp() throws SQLException {
        // Setup queue for this test
        setupQueue(QUEUE_NAME);

        // Create producer
        producer = new Producer(dataSource);
    }


    @Test
    void testProducer() throws SQLException {
        // Act - Produce a message
        producer.produce(QUEUE_NAME, "{\"content\": \"Hello, World!\"}", "{}");

        // Assert - Check if message was produced in PGMQ table
        int messageCount;
        try (var conn = dataSource.getConnection();
             var stmt = conn.createStatement();
             ResultSet rs = stmt.executeQuery("SELECT count(*) FROM pgmq.q_" + QUEUE_NAME)) {
            assertTrue(rs.next(), "Result set should contain data");
            messageCount = rs.getInt(1);
        }

        assertNotEquals(0, messageCount, "Expected to have produced messages, but message count is 0");
    }

    @Test
    void testProducerWithDefaultHeaders() throws SQLException {
        // Act - Produce a message without headers
        producer.produce(QUEUE_NAME, "{\"content\": \"Hello, World 2!\"}");

        // Assert - Check if message was produced
        int messageCount;
        try (Connection conn = dataSource.getConnection();
             Statement stmt = conn.createStatement();
             ResultSet rs = stmt.executeQuery("SELECT count(*) FROM pgmq.q_" + QUEUE_NAME)) {
            assertTrue(rs.next(), "Result set should contain data");
            messageCount = rs.getInt(1);
        }

        assertNotEquals(0, messageCount, "Expected to have produced messages");
    }
}

